""" 
This is the Snakefile that runs the analysis for SARS-CoV-2_chronic_infection_seq.

Author: Will Hannon â€“ https://www.linkedin.com/in/williamhannon/
"""

#### ----------------------- Imports ----------------------- ####

import pandas as pd 
from os.path import join
from snakemake.utils import min_version

min_version('5.6.0')  # require `snakemake` >= 5.6.0

#### -------------------- Configuration -------------------- ####

configfile: "config/config.yml"

#### ------------------------ Utils ------------------------ ####

include: "rules/common.smk"

#### ----------------------- Targets ----------------------- ####

rule all:
    input: 
        join(config['variant_dir'], "variants.csv"),
        join(config['qc_dir'], 'multiqc'),
        join(config['coverage_dir'], "merged.bedgraph"),
        join(config['coverage_dir'], "merged.depth"),
        join(config['coverage_dir'], "merged.average.depth"),
        join(config['coverage_dir'], "coverage.stats")
        
rule clean:
    shell:
        """
        rm -rf logs/
        rm -rf tmp/
        rm -f slurm*.out
        """
#### ------------------------ Rules ------------------------ ####

include: "rules/download-reference.smk"
include: "rules/download-tools.smk"
include: "rules/download-fastq.smk"
include: "rules/index-reference.smk"
include: "rules/trim-adapters.smk"
include: "rules/filter-reads.smk"
include: "rules/align-reads.smk"
include: "rules/calculate-coverage.smk"
include: "rules/calculate-pileup.smk"
include: "rules/qc-tools.smk"
include: "rules/call-variants.smk"
