{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pysam Analysis\n",
    "\n",
    "The goal of this notebook is to vallidate the variant calling and pileup statistics with `pysam`. I can also use pysam to identify local haplotypes by checking if RBD variants are present on the same reads. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "## === Import Libraries === ##\n",
    "import pysam #count variant alleles from BAM\n",
    "import pandas as pd #data frames\n",
    "import numpy as np #arrays\n",
    "import os #interacting with files\n",
    "from Bio import SeqIO #reading fasta format\n",
    "import re #regular expressions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "## === Functions === ##\n",
    "\n",
    "def translate(codon):\n",
    "    \"\"\"\n",
    "    Translate a three letter DNA string into \n",
    "    a one letter amino acid code. \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    codon : str\n",
    "        three letter DNA sequence\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "        one letter amino acid code\n",
    "\n",
    "    Raises\n",
    "    ------\n",
    "    AssertionError\n",
    "        error if codon sequence is invalid\n",
    "        \n",
    "    \"\"\"\n",
    "    \n",
    "    table = { \n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', \n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', \n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', \n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', \n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', \n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', \n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', \n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', \n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', \n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', \n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', \n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', \n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', \n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*', \n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W', \n",
    "    } \n",
    "    \n",
    "    assert codon in table.keys(), \"Not a valid codon sequence.\"\n",
    "    \n",
    "    return table[codon]\n",
    "\n",
    "\n",
    "\n",
    "def mutate(codon, alt, index):\n",
    "    \"\"\"\n",
    "    Replace (mutate) a base in a codon with an \n",
    "    alternate base. \n",
    "        \n",
    "    Parameters\n",
    "    ----------\n",
    "    codon : str\n",
    "        three letter DNA sequence\n",
    "        \n",
    "    alt : str\n",
    "        alternative base\n",
    "        \n",
    "    index : int\n",
    "        index of the alt base in codon (0|1|2). \n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "        codon with alternative base\n",
    "\n",
    "    Raises\n",
    "    ------\n",
    "    AssertionError\n",
    "        error if index is not valid (0|1|2)\n",
    "        \n",
    "    AssertionError\n",
    "        error if base is not valid (A|T|C|G)\n",
    "        \n",
    "    \"\"\"\n",
    "    \n",
    "    assert index in [0,1,2], \"Not a valid index.\"\n",
    "    \n",
    "    assert alt in [\"A\", \"T\", \"C\", \"G\"], \"Not a valid base.\"\n",
    "    \n",
    "    return \"\".join([alt if i == index else b for i,b in enumerate(codon)])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def check_read(read):\n",
    "    \"\"\"\n",
    "    Helper function to decide what reads should\n",
    "    be keep when parsing alignment file with `pysam`. \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    read : AlignedSegment\n",
    "        read from alignment file parsed with `pysam`.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    bool\n",
    "        True/False if read should be included\n",
    "        \n",
    "    \"\"\"\n",
    "    # Exclude Quality Failures\n",
    "    if read.is_qcfail:\n",
    "        return False\n",
    "    # Exclude Secondary Mappings\n",
    "    if read.is_secondary:\n",
    "        return False\n",
    "    # Exclude Unmapped Reads\n",
    "    if read.is_unmapped:\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "def build_af_df(filepath, \n",
    "                callback_function=check_read, \n",
    "                ref = \"NC_045512.2\", \n",
    "                ref_path = \"../../config/ref/SARS2.fa\", \n",
    "                minimum_AF = 0.01, \n",
    "                minimum_qual = 25):\n",
    "    \"\"\"\n",
    "    Read in BAM file and convert to a dataframe containing the frequency of \n",
    "    of any bases present at a given position in the reference genome using the\n",
    "    `pysam` command `count_coverage`. \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    filepath : str\n",
    "        path to the bam file to be parsed\n",
    "        \n",
    "    callback_function : function\n",
    "        function that decides which reads to keep/exclude\n",
    "        \n",
    "    ref : str\n",
    "        name of the contig to count coverage over\n",
    "        \n",
    "    ref_path : str\n",
    "        path to the reference genome as fasta\n",
    "        \n",
    "    minimum_AF : float\n",
    "        minimum allele frequency to include allele in dataframe\n",
    "        \n",
    "    minimum_qual : int\n",
    "        minimum QUAL score to count read at a position.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Pandas.DataFrame\n",
    "       Data Frame containing the bases represented at each positon in the genome\n",
    "        \n",
    "    \"\"\"\n",
    "\n",
    "    # Open alignment with pysam\n",
    "    with pysam.AlignmentFile(filepath, \"rb\") as bamfile:\n",
    "        \n",
    "        # Get a dataframe of the counts\n",
    "        count_df = pd.DataFrame.from_dict({base:counts for base, counts in zip(\"ACGT\", bamfile.count_coverage(contig = ref, read_callback=callback_function, quality_threshold=minimum_qual))})\n",
    "        \n",
    "        # Add the depth at each position\n",
    "        count_df['DP'] = count_df.sum(axis = 1)\n",
    "        \n",
    "        # Add the position \n",
    "        count_df['POS'] = count_df.index + 1\n",
    "        \n",
    "        # Add the reference allele\n",
    "        count_df['REF'] = [base.upper() for base in list(SeqIO.parse(ref_path, \"fasta\"))[0].seq]\n",
    "        \n",
    "        # convert counts to frequency \n",
    "        count_df.iloc[:,0:4] = count_df.iloc[:,0:4].div(count_df.DP, axis = 0)\n",
    "        \n",
    "        # handle any NaNs created by dividing by 0 coverage\n",
    "        count_df = count_df.fillna(0)\n",
    "        \n",
    "        # Melt the data frame to a longer ('tidy') form\n",
    "        count_df = pd.melt(count_df, \n",
    "                           id_vars=['POS', 'DP', 'REF'],\n",
    "                           value_vars=[base for base in 'ATGC'],\n",
    "                           value_name='AF',\n",
    "                           var_name='ALT')\n",
    "        \n",
    "        # Filter out anything less than minimum allele freq\n",
    "        count_df = count_df[count_df['AF'] >= minimum_AF]\n",
    "        \n",
    "        # TRUE/FALSE if it's a SNP\n",
    "        count_df['SNP'] = np.where(count_df['ALT'] != count_df['REF'], True, False)\n",
    "        \n",
    "        # Is a base consensus or not.\n",
    "        count_df['CONS'] = count_df['AF'].map(lambda x: x >= 0.5)\n",
    "    \n",
    "        # Sort by position and reset the index\n",
    "        return count_df.sort_values('POS').reset_index(drop=True)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "def annotate_coding_change_in_spike(count_df, ref_genome):\n",
    "    \"\"\"\n",
    "    Annotate the protein coding changes of mutations in the Spike gene \n",
    "    of SARS-CoV-2 (i.e. in the format `D614G`). \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    count_df : Pandas.DataFrame\n",
    "       Data Frame containing the bases represented at each positon in the genome\n",
    "        \n",
    "    ref_genome : str\n",
    "        reference genome for all of SARS-CoV-2\n",
    "        \n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Pandas.DataFrame\n",
    "       Data Frame containing annotation for all position over Spike\n",
    "        \n",
    "    \"\"\"\n",
    "    # Lists to hold residue changes and positions\n",
    "    residue_change_list = []\n",
    "    residue_position_list = []\n",
    "    resiude_wt_list = []\n",
    "    residue_mut_list = []\n",
    "    \n",
    "    # Spike Sequence by coordinates of spike 0-indexed\n",
    "    spike_sequence = \"\".join(ref_genome[21562:25384])\n",
    "\n",
    "    # Pull out only the SNPs in Spike 1-indexed\n",
    "    ## spike_SNP_df = count_df[(count_df.SNP) & (count_df.POS >= 21563) & (count_df.POS <= 25384)]\n",
    "    \n",
    "    spike_SNP_df = count_df[(count_df.POS >= 21563) & (count_df.POS <= 25384)]\n",
    "\n",
    "    # Iterate over every row in the data frame by alternative allele and position\n",
    "    for alt, pos in zip(spike_SNP_df.ALT.tolist(),spike_SNP_df.POS.tolist()):\n",
    "        \n",
    "        # Get the position in the genome relative to Spike\n",
    "        position_in_spike = pos - 21562\n",
    "        \n",
    "        ## === Check the position of the SNP in the codon == ##\n",
    "\n",
    "        # First base in codon\n",
    "        if position_in_spike % 3 == 1:\n",
    "            \n",
    "            # Save codon position\n",
    "            position_in_codon = 1\n",
    "\n",
    "            # Get the sequence of the wt codon by index\n",
    "            wt_codon = spike_sequence[position_in_spike-1:position_in_spike+2]\n",
    "\n",
    "            # Translate wt codon to residue\n",
    "            wt_aa = translate(wt_codon)\n",
    "\n",
    "            # Get the sequence of the mut codon by index\n",
    "            mut_codon = mutate(wt_codon, alt, position_in_codon-1)\n",
    "\n",
    "            # Translate mut codon to residue\n",
    "            mut_aa = translate(mut_codon)\n",
    "\n",
    "            # Calculate the position of the residue in Spike\n",
    "            residue_position_in_spike = int((position_in_spike + 2) / 3)\n",
    "            \n",
    "            # Build the list of mutations and positions\n",
    "            residue_change_list.append(f\"{wt_aa}{residue_position_in_spike}{mut_aa}\")\n",
    "            residue_position_list.append(residue_position_in_spike)\n",
    "            resiude_wt_list.append(wt_aa)\n",
    "            residue_mut_list.append(mut_aa)\n",
    "\n",
    "\n",
    "        # Second base in codon\n",
    "        elif position_in_spike % 3 == 2:\n",
    "            \n",
    "            # Save codon position\n",
    "            position_in_codon = 2\n",
    "            \n",
    "            # Get the sequence of the wt codon by index\n",
    "            wt_codon = spike_sequence[position_in_spike-2:position_in_spike+1]\n",
    "\n",
    "            # Translate wt codon to residue            \n",
    "            wt_aa = translate(wt_codon)\n",
    "\n",
    "            # Get the sequence of the mut codon by index           \n",
    "            mut_codon = mutate(wt_codon, alt, position_in_codon-1)\n",
    "            \n",
    "            # Translate mut codon to residue\n",
    "            mut_aa = translate(mut_codon)\n",
    "\n",
    "            # Calculate the position of the residue in Spike\n",
    "            residue_position_in_spike = int((position_in_spike + 1) / 3)\n",
    "\n",
    "            # Build the list of mutations and positions\n",
    "            residue_change_list.append(f\"{wt_aa}{residue_position_in_spike}{mut_aa}\")\n",
    "            residue_position_list.append(residue_position_in_spike)\n",
    "            resiude_wt_list.append(wt_aa)\n",
    "            residue_mut_list.append(mut_aa)\n",
    "\n",
    "       \n",
    "        # Third base in codon\n",
    "        elif position_in_spike % 3 == 0:\n",
    "\n",
    "            # Save codon position\n",
    "            position_in_codon = 3\n",
    "            \n",
    "            # Get the sequence of the wt codon by index\n",
    "            wt_codon = spike_sequence[position_in_spike-3:position_in_spike]\n",
    "\n",
    "            # Translate wt codon to residue            \n",
    "            wt_aa = translate(wt_codon)\n",
    "            \n",
    "            # Get the sequence of the mut codon by index  \n",
    "            mut_codon = mutate(wt_codon, alt, position_in_codon-1)\n",
    "\n",
    "            # Translate mut codon to residue\n",
    "            mut_aa = translate(mut_codon)\n",
    "\n",
    "            # Calculate the position of the residue in Spike\n",
    "            residue_position_in_spike = int((position_in_spike) / 3)\n",
    "\n",
    "            # Build the list of mutations and positions\n",
    "            residue_change_list.append(f\"{wt_aa}{residue_position_in_spike}{mut_aa}\")\n",
    "            residue_position_list.append(residue_position_in_spike)\n",
    "            resiude_wt_list.append(wt_aa)\n",
    "            residue_mut_list.append(mut_aa)\n",
    "\n",
    "    return spike_SNP_df.assign(AA_CHANGE = residue_change_list,\n",
    "                               PROT_POS = residue_position_list,\n",
    "                               WT_AA = resiude_wt_list,\n",
    "                               MUT_AA = residue_mut_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ===== Test Functions ===== ##\n",
    "\n",
    "# Path to aligned BAMs\n",
    "filepath = \"../../results/aligned/BWA/\"\n",
    "\n",
    "# Reference genome\n",
    "ref_genome = [base.upper() for base in list(SeqIO.parse(\"../../config/ref/SARS2.fa\", \"fasta\"))[0].seq]\n",
    "\n",
    "# Save the paths to reference files\n",
    "input_bam_paths = []\n",
    "\n",
    "for root, dirs, files in os.walk(filepath):\n",
    "    for file in files:\n",
    "        if file.endswith(\".virus.sorted.marked.bam\"):\n",
    "             input_bam_paths.append(os.path.join(root, file))\n",
    "                \n",
    "# Store the dataframes \n",
    "spike_count_df_list = []\n",
    "\n",
    "# Iterate over paths \n",
    "for path in input_bam_paths:\n",
    "    \n",
    "    # Name of sample\n",
    "    accession = os.path.basename(path)\n",
    "    \n",
    "    # Get the day of the sample \n",
    "    regex = re.compile(\"Day(?P<Day>\\d+)_\")\n",
    "    m = regex.match(accession)\n",
    "    day = int(m.group('Day'))\n",
    "    \n",
    "    # Build the count df\n",
    "    count_df = build_af_df(path, \n",
    "                callback_function=check_read, \n",
    "                ref = \"NC_045512.2\", \n",
    "                ref_path = \"../../config/ref/SARS2.fa\", \n",
    "                minimum_AF = 0.01, \n",
    "                minimum_qual = 25)\n",
    "    \n",
    "    # Add accession and day to data frame\n",
    "    count_df.insert(0, 'ACCESSION', accession)\n",
    "    count_df.insert(1, 'DAY', day)\n",
    "    \n",
    "    # Annotate the count df   \n",
    "    spike_count_df = annotate_coding_change_in_spike(count_df, ref_genome)\n",
    "\n",
    "    # Save the list of dfs \n",
    "    spike_count_df_list.append(spike_count_df)\n",
    "\n",
    "# Join the list of dataframe\n",
    "joined_spike_count_df = pd.concat(spike_count_df_list, ignore_index=True)\n",
    "\n",
    "# TRUE/FALSE if there is a missense mutation\n",
    "joined_spike_count_df['MISSENSE'] = np.where(joined_spike_count_df['WT_AA'] != joined_spike_count_df['MUT_AA'], True, False)\n",
    "\n",
    "# TRUE/FALSE if the mutation is in the RBD\n",
    "joined_spike_count_df['RBD'] = np.where((joined_spike_count_df['PROT_POS'] >= 331) & (joined_spike_count_df['PROT_POS'] <= 531), True, False)\n",
    "\n",
    "# Write this dataframe into R for temporary plotting and testing.\n",
    "joined_spike_count_df.to_csv(\"../../config/pysam_pileup.csv\",index=False)\n",
    "\n",
    "## ===== Test Functions ===== ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ACCESSION</th>\n",
       "      <th>DAY</th>\n",
       "      <th>POS</th>\n",
       "      <th>DP</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CONS</th>\n",
       "      <th>AA_CHANGE</th>\n",
       "      <th>PROT_POS</th>\n",
       "      <th>WT_AA</th>\n",
       "      <th>MUT_AA</th>\n",
       "      <th>MISSENSE</th>\n",
       "      <th>RBD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1320</th>\n",
       "      <td>Day18_Fullgenome_5_N_4_28_20frg_321634w_NA11.B...</td>\n",
       "      <td>18</td>\n",
       "      <td>22880</td>\n",
       "      <td>704</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>N440N</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1479</th>\n",
       "      <td>Day18_Fullgenome_5_N_4_28_20frg_321634w_NA11.B...</td>\n",
       "      <td>18</td>\n",
       "      <td>23039</td>\n",
       "      <td>1381</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493Q</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24423</th>\n",
       "      <td>Day25_Fullgenome_6_N_5_5_20frg_321634w_NA12.BW...</td>\n",
       "      <td>25</td>\n",
       "      <td>23039</td>\n",
       "      <td>880</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493Q</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24264</th>\n",
       "      <td>Day25_Fullgenome_6_N_5_5_20frg_321634w_NA12.BW...</td>\n",
       "      <td>25</td>\n",
       "      <td>22880</td>\n",
       "      <td>516</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>N440N</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5140</th>\n",
       "      <td>Day75_Fullgenome_8_N_6_24_20frg_321634w_NB2.BW...</td>\n",
       "      <td>75</td>\n",
       "      <td>22880</td>\n",
       "      <td>420</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>N440N</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5300</th>\n",
       "      <td>Day75_Fullgenome_8_N_6_24_20frg_321634w_NB2.BW...</td>\n",
       "      <td>75</td>\n",
       "      <td>23039</td>\n",
       "      <td>1221</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493Q</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31916</th>\n",
       "      <td>Day81_Fullgenome_7_N_6_30_20frg_321634w_NB1.BW...</td>\n",
       "      <td>81</td>\n",
       "      <td>22880</td>\n",
       "      <td>419</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>N440N</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32075</th>\n",
       "      <td>Day81_Fullgenome_7_N_6_30_20frg_321634w_NB1.BW...</td>\n",
       "      <td>81</td>\n",
       "      <td>23039</td>\n",
       "      <td>787</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493Q</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12953</th>\n",
       "      <td>Day128_Fullgenome_A5_20frg_819_NP_08_16_321948...</td>\n",
       "      <td>128</td>\n",
       "      <td>23039</td>\n",
       "      <td>1095</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493K</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>K</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12794</th>\n",
       "      <td>Day128_Fullgenome_A5_20frg_819_NP_08_16_321948...</td>\n",
       "      <td>128</td>\n",
       "      <td>22880</td>\n",
       "      <td>1143</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>N440N</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28092</th>\n",
       "      <td>Day130_Fullgenome_A6_20frg_819_NP_08_18_321948...</td>\n",
       "      <td>130</td>\n",
       "      <td>22880</td>\n",
       "      <td>743</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.635262</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>N440N</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28093</th>\n",
       "      <td>Day130_Fullgenome_A6_20frg_819_NP_08_18_321948...</td>\n",
       "      <td>130</td>\n",
       "      <td>22880</td>\n",
       "      <td>743</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.364738</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>N440D</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28252</th>\n",
       "      <td>Day130_Fullgenome_A6_20frg_819_NP_08_18_321948...</td>\n",
       "      <td>130</td>\n",
       "      <td>23039</td>\n",
       "      <td>1226</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.999184</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493K</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>K</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20426</th>\n",
       "      <td>Day143_Fullgenome_A2_20frg_819_8_31_322026w_AG...</td>\n",
       "      <td>143</td>\n",
       "      <td>22880</td>\n",
       "      <td>196</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>N440N</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20589</th>\n",
       "      <td>Day143_Fullgenome_A2_20frg_819_8_31_322026w_AG...</td>\n",
       "      <td>143</td>\n",
       "      <td>23039</td>\n",
       "      <td>1531</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493Q</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16757</th>\n",
       "      <td>Day146_Fullgenome_A10_20frg_819_NP_9_3_20_3220...</td>\n",
       "      <td>146</td>\n",
       "      <td>23039</td>\n",
       "      <td>864</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.998843</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493K</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>K</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16597</th>\n",
       "      <td>Day146_Fullgenome_A10_20frg_819_NP_9_3_20_3220...</td>\n",
       "      <td>146</td>\n",
       "      <td>22880</td>\n",
       "      <td>348</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>N440D</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9126</th>\n",
       "      <td>Day152_Fullgenome_7_20frg_819_N3_9_9_322214w_A...</td>\n",
       "      <td>152</td>\n",
       "      <td>23039</td>\n",
       "      <td>1330</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.378947</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Q493K</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>K</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9125</th>\n",
       "      <td>Day152_Fullgenome_7_20frg_819_N3_9_9_322214w_A...</td>\n",
       "      <td>152</td>\n",
       "      <td>23039</td>\n",
       "      <td>1330</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.621053</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Q493Q</td>\n",
       "      <td>493</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8963</th>\n",
       "      <td>Day152_Fullgenome_7_20frg_819_N3_9_9_322214w_A...</td>\n",
       "      <td>152</td>\n",
       "      <td>22880</td>\n",
       "      <td>634</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>N440N</td>\n",
       "      <td>440</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               ACCESSION  DAY    POS    DP  \\\n",
       "1320   Day18_Fullgenome_5_N_4_28_20frg_321634w_NA11.B...   18  22880   704   \n",
       "1479   Day18_Fullgenome_5_N_4_28_20frg_321634w_NA11.B...   18  23039  1381   \n",
       "24423  Day25_Fullgenome_6_N_5_5_20frg_321634w_NA12.BW...   25  23039   880   \n",
       "24264  Day25_Fullgenome_6_N_5_5_20frg_321634w_NA12.BW...   25  22880   516   \n",
       "5140   Day75_Fullgenome_8_N_6_24_20frg_321634w_NB2.BW...   75  22880   420   \n",
       "5300   Day75_Fullgenome_8_N_6_24_20frg_321634w_NB2.BW...   75  23039  1221   \n",
       "31916  Day81_Fullgenome_7_N_6_30_20frg_321634w_NB1.BW...   81  22880   419   \n",
       "32075  Day81_Fullgenome_7_N_6_30_20frg_321634w_NB1.BW...   81  23039   787   \n",
       "12953  Day128_Fullgenome_A5_20frg_819_NP_08_16_321948...  128  23039  1095   \n",
       "12794  Day128_Fullgenome_A5_20frg_819_NP_08_16_321948...  128  22880  1143   \n",
       "28092  Day130_Fullgenome_A6_20frg_819_NP_08_18_321948...  130  22880   743   \n",
       "28093  Day130_Fullgenome_A6_20frg_819_NP_08_18_321948...  130  22880   743   \n",
       "28252  Day130_Fullgenome_A6_20frg_819_NP_08_18_321948...  130  23039  1226   \n",
       "20426  Day143_Fullgenome_A2_20frg_819_8_31_322026w_AG...  143  22880   196   \n",
       "20589  Day143_Fullgenome_A2_20frg_819_8_31_322026w_AG...  143  23039  1531   \n",
       "16757  Day146_Fullgenome_A10_20frg_819_NP_9_3_20_3220...  146  23039   864   \n",
       "16597  Day146_Fullgenome_A10_20frg_819_NP_9_3_20_3220...  146  22880   348   \n",
       "9126   Day152_Fullgenome_7_20frg_819_N3_9_9_322214w_A...  152  23039  1330   \n",
       "9125   Day152_Fullgenome_7_20frg_819_N3_9_9_322214w_A...  152  23039  1330   \n",
       "8963   Day152_Fullgenome_7_20frg_819_N3_9_9_322214w_A...  152  22880   634   \n",
       "\n",
       "      REF ALT        AF    SNP   CONS AA_CHANGE  PROT_POS WT_AA MUT_AA  \\\n",
       "1320    A   A  1.000000  False   True     N440N       440     N      N   \n",
       "1479    C   C  1.000000  False   True     Q493Q       493     Q      Q   \n",
       "24423   C   C  1.000000  False   True     Q493Q       493     Q      Q   \n",
       "24264   A   A  1.000000  False   True     N440N       440     N      N   \n",
       "5140    A   A  1.000000  False   True     N440N       440     N      N   \n",
       "5300    C   C  1.000000  False   True     Q493Q       493     Q      Q   \n",
       "31916   A   A  1.000000  False   True     N440N       440     N      N   \n",
       "32075   C   C  1.000000  False   True     Q493Q       493     Q      Q   \n",
       "12953   C   A  1.000000   True   True     Q493K       493     Q      K   \n",
       "12794   A   A  1.000000  False   True     N440N       440     N      N   \n",
       "28092   A   A  0.635262  False   True     N440N       440     N      N   \n",
       "28093   A   G  0.364738   True  False     N440D       440     N      D   \n",
       "28252   C   A  0.999184   True   True     Q493K       493     Q      K   \n",
       "20426   A   A  1.000000  False   True     N440N       440     N      N   \n",
       "20589   C   C  1.000000  False   True     Q493Q       493     Q      Q   \n",
       "16757   C   A  0.998843   True   True     Q493K       493     Q      K   \n",
       "16597   A   G  1.000000   True   True     N440D       440     N      D   \n",
       "9126    C   A  0.378947   True  False     Q493K       493     Q      K   \n",
       "9125    C   C  0.621053  False   True     Q493Q       493     Q      Q   \n",
       "8963    A   A  1.000000  False   True     N440N       440     N      N   \n",
       "\n",
       "       MISSENSE   RBD  \n",
       "1320      False  True  \n",
       "1479      False  True  \n",
       "24423     False  True  \n",
       "24264     False  True  \n",
       "5140      False  True  \n",
       "5300      False  True  \n",
       "31916     False  True  \n",
       "32075     False  True  \n",
       "12953      True  True  \n",
       "12794     False  True  \n",
       "28092     False  True  \n",
       "28093      True  True  \n",
       "28252      True  True  \n",
       "20426     False  True  \n",
       "20589     False  True  \n",
       "16757      True  True  \n",
       "16597      True  True  \n",
       "9126       True  True  \n",
       "9125      False  True  \n",
       "8963      False  True  "
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Positions to check: \n",
    "#Q493K = 23039\n",
    "#N440D = 22880\n",
    "\n",
    "joined_spike_count_df.loc[joined_spike_count_df['POS'].isin([23039, 22880])].sort_values('DAY')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
