---
title: "SARS-CoV-2 Chronic Infection"
author: "Will Hannon"
date: "11/20/2020"
output: html_document 
---

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```


```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed -- non-BioManager installed
packages = c("tidyverse", "seqinr")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```


```{r Filepaths, include=FALSE}
## ==== Important file paths ==== ##
pysam.data = "../../results/pileup/pysam_variants.csv" # pileup statistics from pysam
samtools.data = "../../results/pileup/samtools_variants.csv" # pileup statistics from samtools
depth.data = "../../results/coverage/merged.depth" # coverage statistics from samtools
variant.data = "../../results/variant/variants.csv" # varaints from lofreq and varscan
sample.data = "../../config/samples.csv" # sample data used to run the analysis

## ==== Paths to primer locataions and insert locations ==== ##
insert.bed = "../../config/nCoV-2019.insert.bed.txt"
primers.bed = "../../config/nCoV-2019.primer.bed.txt"

## ==== Path to Reference genome to annotate the coding changes ==== ##
ref.path = "../../config/ref/SARS2.fa"

```

```{r Coding Change Function, message=F, warning=F, echo=F}

## === Function to annotate the coding change based on position in Spike === ##

# only works in Spike for SARS-CoV-2 #
annotate.coding.change = function(POS, ALT, ref.genome) {
  
  #' Function takes as argument the nucleotide positon in the
  #' genome (1-based indexing) and the alternative allele, and 
  #' returns the coordinates of the residue in Spike as well 
  #' as the the residue change if the mutation is non-synonymous.
  
  # Extract the sequence of Spike from a fasta file
  spike.seq = ref.genome$NC_045512.2[21563:25384]
  
  # Convert the positon in the genome to the positon in Spike 
  POS.spike = POS - 21562
  
  # Get position in the codon by the Spike index 
  if (POS.spike %% 3 == 1) {
    
    # Start of the codon 
    codon.pos = 1
    # Get the codon
    codon.seq = spike.seq[(POS.spike):(POS.spike+2)]
    # Translate
    codon.aa = translate(codon.seq, frame = 0, sens = "F")
    # Change the seq to alt
    codon.seq[codon.pos] = tolower(ALT)
    # Translate the alternative codon
    codon.aa.alt = translate(codon.seq, frame = 0, sens = "F")
    # Return the change
    return(paste0(codon.aa, ((POS.spike+2)/3), codon.aa.alt))
    
  } else if (POS.spike %% 3 == 2) {
    
    # Middle of the codon 
    codon.pos = 2
    # Get the codon
    codon.seq = spike.seq[(POS.spike-1):(POS.spike+1)]
    # Translate
    codon.aa = translate(codon.seq, frame = 0, sens = "F")
    # Change the seq to alt
    codon.seq[codon.pos] = tolower(ALT)
    # Translate the alternative codon
    codon.aa.alt = translate(codon.seq, frame = 0, sens = "F")
    # Return the change
    return(paste0(codon.aa, ((POS.spike+1)/3), codon.aa.alt))
    
  } else if (POS.spike %% 3 == 0) {
    
    # End of the codon 
    codon.pos = 3
    # Get the codon
    codon.seq = spike.seq[(POS.spike-2):(POS.spike)]
    # Translate
    codon.aa = translate(codon.seq, frame = 0, sens = "F")
    # Change the seq to alt
    codon.seq[codon.pos] = tolower(ALT)
    # Translate the alternative codon
    codon.aa.alt = translate(codon.seq, frame = 0, sens = "F")
    # Return the change
    return(paste0(codon.aa, ((POS.spike)/3), codon.aa.alt))
  
  }

}

```

## Summary

This notebook contains the code for vallidating the main figures associated with the longitudinal deep sequencing analysis of SARS-CoV-2 from an immunocompromised patient. The associated publication for these samples is located [here](https://www.nejm.org/doi/full/10.1056/NEJMc2031364).

Runs containing raw, paired-end reads, for each of nine timepoints since initial infection (days 18, 25, 75, 81, 128, 130, 143, 146, and 152) were filtered and trimmed using `fastp`. Reads matching SARS-CoV-2 sequences were filtered using kmer matching with `BBDuk`. These reads were aligned to the Wuhan-Hu-1 reference genome (NC_045512.2) with `BWA.` SNPs were identified in each sample with the variant caller `lofreq` and confirmed by parsing the read pileup statistics generated by `samtools mpileup`. A quality score cutoff of 25 was used to identify variants.

## Coverage 

Below, I plotted the coverage of reads over the Spike protein. Coverage was calculated using `samtools depth` with a quality score cutoff of PHRED = 25.

```{r Coverage, message=F, warning=F, echo=F, fig.width=10, fig.height=6}

# == Define the regions of interst - Spike, RBD, and RBM == #
spike.coords = c(21563, 25384)
rbd.coords = c(22556, 23158)
rbm.coords = c(22871, 23086)

# == Import the coverage data == #
# Calculate the number of base observavtions in 10bp bins
samtools.depth.bins = left_join(read.table(depth.data, header = T), read_csv(sample.data), by = c("Accession" = "Run"))

# Plot the depth over Spike - no cuttoff
samtools.depth.bins %>% 
  ggplot(aes(x = (Start+Stop)/2, y = Depth, col = as.factor(Day))) + 
    geom_line(size = 1.5) +
    scale_color_viridis_d() + 
    scale_x_continuous(limits = c(spike.coords[1], spike.coords[2])) + 
    xlab("Position") + 
    ylab("Depth") + 
    annotate("rect", xmin = spike.coords[1], xmax = spike.coords[2], ymin = -100, ymax = -1000, fill = "darkgoldenrod1", col = "black", alpha = 0.75) + 
    annotate("text", x = (spike.coords[1] + spike.coords[2])/2, y = (-100 + -1000)/2, label = "Spike") + 
    annotate("rect", xmin = rbd.coords[1], xmax = rbd.coords[2], ymin = -1100, ymax = -2000, fill = "coral2", col = "black", alpha = 0.75) + 
    annotate("text", x = (rbd.coords[1] + rbd.coords[2])/2, y = (-1100 + -2000)/2, label = "RBD") + 
    annotate("rect", xmin = rbm.coords[1], xmax = rbm.coords[2], ymin = -2100, ymax = -3000, fill = "cadetblue2", col = "black", alpha = 0.75) +
    annotate("text", x = (rbm.coords[1] + rbm.coords[2])/2, y = (-2100 + -3000)/2, label = "RBM") +
    theme_classic() +
    theme(legend.position="none") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 

# Save the file to figures directory
# ggsave(paste0("../../config/figures/", format(Sys.time(), "%Y-%m-%d_H%IM%M"), "_", "Coverage_Over_Spike.png") , width = 10, height = 6, dpi = 300)
```

To get a clearer picture of coverage, I'm capping the max coverage at 500X. 

```{r Coverage Max, message=F, warning=F, echo=F, fig.width=10, fig.height=6}

# Plot the depth over Spike - no cuttoff
samtools.depth.bins %>% 
  mutate(Depth = ifelse(Depth <= 500, Depth, 500)) %>%
  ggplot(aes(x = (Start+Stop)/2, y = Depth, col = as.factor(Day))) + 
    geom_line(size = 1.5) +
    geom_hline(yintercept = 100, size = 1.5, linetype = 2, col = "red") +
    scale_color_viridis_d() + 
    scale_y_continuous(breaks = c(0, 100, 200, 300, 400, 500), labels = c("0", "100", "200", "300", "400", expression(phantom(x) >=500))) +
    scale_x_continuous(limits = c(spike.coords[1], spike.coords[2])) + 
    xlab("Position") + 
    ylab("Depth") + 
    annotate("rect", xmin = spike.coords[1], xmax = spike.coords[2], ymin = -10, ymax = -50, fill = "darkgoldenrod1", col = "black", alpha = 0.75) + 
    annotate("text", x = (spike.coords[1] + spike.coords[2])/2, y = (-5 + -50)/2, label = "Spike") + 
    annotate("rect", xmin = rbd.coords[1], xmax = rbd.coords[2], ymin = -55, ymax = -100, fill = "coral2", col = "black", alpha = 0.75) + 
    annotate("text", x = (rbd.coords[1] + rbd.coords[2])/2, y = (-55 + -100)/2, label = "RBD") + 
    annotate("rect", xmin = rbm.coords[1], xmax = rbm.coords[2], ymin = -105, ymax = -150, fill = "cadetblue2", col = "black", alpha = 0.75) +
    annotate("text", x = (rbm.coords[1] + rbm.coords[2])/2, y = (-105 + -150)/2, label = "RBM") +
    theme_classic() +
    theme(legend.position="none") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 


# Save the file to figures directory
# ggsave(paste0("../../config/figures/", format(Sys.time(), "%Y-%m-%d_H%IM%M"), "_", "Coverage_Over_Spike_500X_Max.png") , width = 10, height = 6, dpi = 300)
```

## Single Nucleotide Polymorphisms in Spike. 

```{r Process Variants, message=F, warning=F, echo=F}

## == Process the variant dataframe from lofreq and varscan and annotate == ##

# Define the reference genome
ref.genome = read.fasta(ref.path)

# Import data
variant.df = read_csv(variant.data) %>% 
  # Get only the variants in the Spike gene
  filter(POS >= spike.coords[1] & POS <= spike.coords[2]) 

# Determine what the AA changes are 
aa_changes = c()

for (i in (1:nrow(variant.df))) {
  
  # Get the position and the alt allele
  position = variant.df[i,]$POS
  alternative = variant.df[i,]$ALT
  
  # Determine the amino acid change
  aa_change = annotate.coding.change(POS=position, ALT = alternative, ref.genome)
  
  # Append the aa changes to vector
  aa_changes = append(aa_changes, aa_change)
  
}
 
# Append the AA change to the df
variant.df$AA_change = aa_changes

# Get protein position, AA ref, and AA alt
variant.df = variant.df %>% 
  mutate(AA_POS = as.numeric(str_extract(AA_change, "[[:digit:]]+"))) %>% 
  separate(AA_change, 
           into = c("AA_REF", "AA_ALT"), 
           sep = "[[:digit:]]+", remove = FALSE) %>% 
  mutate(SNP = paste0(REF, POS, ALT))

```

Below, I plot all of the mutations in Spike at any timepoint if they are above 1% frequency in both `lofreq` and `varscan`. Some of the lower frequency mutations are only found in `lofreq`. The major cluster of mutations in the RBD/RBM.
```{r Variants, message=F, warning=F, echo=F, fig.width=15, fig.height=5}

variant.df %>% 
  # Filter for minimum AF and Spike coordinates
  filter(AF >= 0.01, POS >= spike.coords[1] & POS <= spike.coords[2]) %>% 
  # Plot the mutations in Spike
  ggplot(aes(x = POS, y = AF, col = as.factor(Day), fill = as.factor(Day))) +
    geom_point()  +
    geom_bar(stat="identity", width=.01, position = position_identity()) +    
    facet_wrap(~Caller) +
    scale_color_viridis_d() + 
    scale_fill_viridis_d() + 
    scale_x_continuous(limits = c(spike.coords[1], spike.coords[2])) + 
    xlab("Position") + 
    ylab("Depth") + 
    labs(color='Day', fill='Day') +
    annotate("rect", xmin = spike.coords[1], xmax = spike.coords[2], ymin = -.001, ymax = -.05, fill = "darkgoldenrod1", col = "black", alpha = 0.75) + 
    annotate("text", x = (spike.coords[1] + spike.coords[2])/2, y = (-.001 + -.05)/2, label = "Spike") + 
    annotate("rect", xmin = rbd.coords[1], xmax = rbd.coords[2], ymin = -.06, ymax = -.11, fill = "coral2", col = "black", alpha = 0.75) + 
    annotate("text", x = (rbd.coords[1] + rbd.coords[2])/2, y = (-.06 + -.11)/2, label = "RBD") + 
    annotate("rect", xmin = rbm.coords[1], xmax = rbm.coords[2], ymin = -.12, ymax = -.17, fill = "cadetblue2", col = "black", alpha = 0.75) +
    annotate("text", x = (rbm.coords[1] + rbm.coords[2])/2, y = (-.12 + -.17)/2, label = "RBM") +
    theme_classic() +
    theme(legend.position="right") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=18,  family="Helvetica")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) 
  

# Save the file to figures directory
# ggsave(paste0("../../config/figures/", format(Sys.time(), "%Y-%m-%d_H%IM%M"), "_", "lofreq_variants_over_spike.png") , width = 10, height = 6, dpi = 300)

```


## Comparison between `lofreq`, `varscan`, and `pileup` variants.

*Filtering criteria:*

- Mutation must rise above 10% at some point during sampling. 

- Mutation must be a missense mutation. 

- Mutation must change in frequency over the time interval of interest. 

- Call any other mutations at the same protein position. 

First, showing all mutations in Spike: 

```{r Variant Timecourse Spike - lofreq, message=F, warning=F, echo=F, fig.width=15, fig.height=8}

## ==== Add matplotlib Tableau color scheme ==== ##
tab_colors = c("#1f77b4", "#ff7f0e")

## ==== Get a list of the amino acid positions to plot based on filtering criteria ==== ##
variant.sites.to.plot = variant.df %>% 
  # Must have an AF greater than 10%
  filter(AF >= 0.1) %>% 
  # Must be a missense mutation
  mutate(MISSENSE = ifelse(AA_REF != AA_ALT, "TRUE", "FALSE")) %>% 
  filter(MISSENSE == "TRUE") %>% 
  # Get a list of the relevant protein positions
  pull(AA_POS) %>% 
  # only the unqiue sites are necessary
  unique()

## ==== Further filtering to remove sites that don't change over the interval ==== ##
variant.timecouse.df = variant.df[which(variant.df$AA_POS %in% variant.sites.to.plot),] %>% 
  # Select only the necessary columns
  select(AA_POS, POS, REF, ALT, AA_change, AF, Day, Caller, AA_REF, AA_ALT) %>% 
  # Make a new column for pivot 
  mutate(Day = paste0("DAY_", Day)) %>% 
  # Pivot wider to have information for every single time point
  pivot_wider(names_from = Day, values_from = AF, values_fill = 0) %>%  
  # Pivot longer to tidy version. 
  pivot_longer(starts_with("DAY_"), names_to = "Day", values_to = "AF") %>% 
  # Split the day back into a number 
  mutate(Day = as.integer(str_replace_all(Day, "DAY_", ""))) 
 
## ==== Plotting the variant timecouse ==== ##
y.pos = 0
point.size = 1
adj = 0.05
tick.color = "#bcbd22"

variant.timecouse.df %>% 
  # Remove D614G which is present in all timepoints.
  filter(AA_change != "D614G") %>%
  # Add "residue" to the name for plotting
  mutate(RESIDUE = fct_reorder(paste("Residue", AA_POS, sep = " "), AA_POS, .desc = F)) %>% 
  
  filter(Caller == "lofreq") %>% 
  ## ========= Plot the figure ========= ##
  
  ggplot(aes(x = Day, y = AF, group = AA_ALT)) +
    # Annotation to represent the WT allele frequency
    annotate(geom = "rect", xmin = 18, xmax = 152, ymin=0, ymax=1, fill = "#7f7f7f") +
    geom_area(fill = "#ff7f0e") +
    #geom_point(col ="black") +
    facet_wrap(~RESIDUE, ncol = 6) +
    ggtitle("Lofeq") + 
    ylab("mutation frequency") + 
    xlab("days after diagnosis") + 
    scale_fill_manual(values = tab_colors) +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(-.05,1) , expand = c(0,0)) +
    scale_x_continuous(limits = c(18,152) , expand = c(0,0)) +
    ### ==== Segments/Point Annotations X-Axis ==== ###
    annotate(geom = "segment", x = 145, y = 0 ,  xend = 145, yend = 1, size = 1, col = "black", linetype = 4) +
    annotate(geom = "segment", x = 18, y = 0 ,  xend = 152, yend = 0, size = 1, col = "black", linetype = 1) +
  
    annotate(geom = "segment", x = 18, y = y.pos ,  xend = 18, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 25, y = y.pos ,  xend = 25, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 75, y = y.pos ,  xend = 75, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 81, y = y.pos ,  xend = 81, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 128, y = y.pos ,  xend = 128, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 130, y = y.pos ,  xend = 130, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 143, y = y.pos ,  xend = 143, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 146, y = y.pos ,  xend = 146, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 152, y = y.pos ,  xend = 152, yend = y.pos+adj, size = 1, col = tick.color) +
    ### ==== Segments/Point Annotations X-Axis ==== ###
    
    ### ==== Themes for plot aestetics ==== ###
  
  
    theme_classic() +
    theme(
      # Set the appropriate font size
      text=element_text(size=20,  family="Helvetica"),
      # Change x axis title position
      axis.title.x = element_text(vjust= -0.5),
      # Change y axis title position
      axis.title.y = element_text(vjust= 1),
      # Change the distace of lables from x axis
      axis.text.x = element_text(vjust= -1),
      # Change the distace of lables from y axis
      axis.text.y = element_text(hjust= 2),
      # Remove the legend
      legend.position = "none",
      # Change the margin to make sure x axis labels aren't cut-off
      plot.margin = margin(10, 30, 10, 10)
    ) + 
    theme(plot.title = element_text(hjust = 0.5))
  

```


```{r Variant Timecourse Spike - varscan, message=F, warning=F, echo=F, fig.width=15, fig.height=8}

## ==== Plotting the variant timecouse ==== ##
y.pos = 0
point.size = 1
adj = 0.05
tick.color = "#bcbd22"

variant.timecouse.df %>% 
  # Remove D614G which is present in all timepoints.
  filter(AA_change != "D614G") %>%
  # Add "residue" to the name for plotting
  mutate(RESIDUE = fct_reorder(paste("Residue", AA_POS, sep = " "), AA_POS, .desc = F)) %>% 
  
  filter(Caller == "varscan") %>% 
  ## ========= Plot the figure ========= ##
  
  ggplot(aes(x = Day, y = AF, group = AA_ALT)) +
    # Annotation to represent the WT allele frequency
    annotate(geom = "rect", xmin = 18, xmax = 152, ymin=0, ymax=1, fill = "#7f7f7f") +
    geom_area(fill = "#ff7f0e") +
    #geom_point(col ="black") +
    facet_wrap(~RESIDUE, ncol = 6) +
    ggtitle("Varscan") + 
    ylab("mutation frequency") + 
    xlab("days after diagnosis") + 
    scale_fill_manual(values = tab_colors) +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(-.05,1) , expand = c(0,0)) +
    scale_x_continuous(limits = c(18,152) , expand = c(0,0)) +
    ### ==== Segments/Point Annotations X-Axis ==== ###
    annotate(geom = "segment", x = 145, y = 0 ,  xend = 145, yend = 1, size = 1, col = "black", linetype = 4) +
    annotate(geom = "segment", x = 18, y = 0 ,  xend = 152, yend = 0, size = 1, col = "black", linetype = 1) +
  
    annotate(geom = "segment", x = 18, y = y.pos ,  xend = 18, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 25, y = y.pos ,  xend = 25, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 75, y = y.pos ,  xend = 75, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 81, y = y.pos ,  xend = 81, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 128, y = y.pos ,  xend = 128, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 130, y = y.pos ,  xend = 130, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 143, y = y.pos ,  xend = 143, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 146, y = y.pos ,  xend = 146, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 152, y = y.pos ,  xend = 152, yend = y.pos+adj, size = 1, col = tick.color) +
    ### ==== Segments/Point Annotations X-Axis ==== ###
    
    ### ==== Themes for plot aestetics ==== ###
  
  
    theme_classic() +
    theme(
      # Set the appropriate font size
      text=element_text(size=20,  family="Helvetica"),
      # Change x axis title position
      axis.title.x = element_text(vjust= -0.5),
      # Change y axis title position
      axis.title.y = element_text(vjust= 1),
      # Change the distace of lables from x axis
      axis.text.x = element_text(vjust= -1),
      # Change the distace of lables from y axis
      axis.text.y = element_text(hjust= 2),
      # Remove the legend
      legend.position = "none",
      # Change the margin to make sure x axis labels aren't cut-off
      plot.margin = margin(10, 30, 10, 10)
    ) +
    theme(plot.title = element_text(hjust = 0.5))
  

```

```{r Variant Timecourse Spike - pysam, message=F, warning=F, echo=F, fig.width=15, fig.height=8}
## ==== Add matplotlib Tableau color scheme ==== ##
tab_colors = c("#1f77b4","#ff7f0e")

## ==== Read in the pysam/pileup data ==== ##
pysam.df = read_csv(pysam.data)

## ==== Get a list of the amino acid positions to plot ==== ##
sites.to.plot = pysam.df %>% 
  # Must have at least one missense mutation
  filter(MISSENSE == "TRUE") %>% 
  # At least one mutation over 10%
  filter(AF >= 0.1) %>% 
  # Get a list of the relevant protein positions
  pull(PROT_POS) %>% 
  # only the unqiue sites are necessary
  unique()

## ==== Further filtering to remove sites that don't change over the interval ==== ##
figure.df = pysam.df[which(pysam.df$PROT_POS %in% sites.to.plot),] %>% 
  # Get only the missense mutations
  filter(MISSENSE == "TRUE") %>% 
  # Drop the extra columns 
  select(!c("ACCESSION", "MISSENSE", "DP", "CONS", "SNP")) %>% 
  # Make a new column for pivot 
  mutate(DAY = paste0("DAY_", DAY)) %>% 
  # Pivot wider to have information for every single time point
  pivot_wider(names_from = DAY, values_from = AF, values_fill = 0) %>%  
  # Pivot longer to tidy version. 
  pivot_longer(starts_with("DAY_"), names_to = "DAY", values_to = "AF") %>% 
  # Split the day back into a number 
  mutate(DAY = as.integer(str_replace_all(DAY, "DAY_", ""))) %>% 
  # Add the color information for plotting
  mutate(COLOR_SCHEME = case_when(AA_CHANGE == "N440D" ~ "ESCAPE",
                                  AA_CHANGE == "F486I" ~ "ESCAPE",
                                  AA_CHANGE == "Q493K" ~ "ESCAPE",
                                  AA_CHANGE == "E484A" ~ "ESCAPE",
                                  AA_CHANGE == "Y489H" ~ "ESCAPE",
                                  TRUE ~ "NON-ESCAPE"))


## ==== Plot Figure ==== ##
y.pos = 0
point.size = 1
adj = 0.05
tick.color = "#bcbd22"

figure.df %>% 
  # Remove D614G which is present in all timepoints.
  filter(AA_CHANGE != "D614G") %>%
  # Add "residue" to the name for plotting
  mutate(RESIDUE = fct_reorder(paste("Residue", PROT_POS, sep = " "), PROT_POS, .desc = F)) %>% 
  ## ========= Plot the figure ========= ##
  
  ggplot(aes(x = DAY, y = AF, fill = COLOR_SCHEME, group = MUT_AA)) +
    # Annotation to represent the WT allele frequency
    annotate(geom = "rect", xmin = 18, xmax = 152, ymin=0, ymax=1, fill = "#7f7f7f") +
    geom_area() +
    #geom_point(col ="black") +
    facet_wrap(~RESIDUE, ncol = 6) +
    ylab("mutation frequency") + 
    xlab("days after diagnosis") + 
    ggtitle("Psyam") +
    scale_fill_manual(values = tab_colors) +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(-.05,1) , expand = c(0,0)) +
    scale_x_continuous(limits = c(18,152) , expand = c(0,0)) +
    ### ==== Segments/Point Annotations X-Axis ==== ###
    annotate(geom = "segment", x = 145, y = 0 ,  xend = 145, yend = 1, size = 1, col = "black", linetype = 4) +
    annotate(geom = "segment", x = 18, y = 0 ,  xend = 152, yend = 0, size = 1, col = "black", linetype = 1) +
  
    annotate(geom = "segment", x = 18, y = y.pos ,  xend = 18, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 25, y = y.pos ,  xend = 25, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 75, y = y.pos ,  xend = 75, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 81, y = y.pos ,  xend = 81, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 128, y = y.pos ,  xend = 128, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 130, y = y.pos ,  xend = 130, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 143, y = y.pos ,  xend = 143, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 146, y = y.pos ,  xend = 146, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 152, y = y.pos ,  xend = 152, yend = y.pos+adj, size = 1, col = tick.color) +
    ### ==== Segments/Point Annotations X-Axis ==== ###
    
    ### ==== Themes for plot aestetics ==== ###
  
  
    theme_classic() +
    theme(
      # Set the appropriate font size
      text=element_text(size=20,  family="Helvetica"),
      # Change x axis title position
      axis.title.x = element_text(vjust= -0.5),
      # Change y axis title position
      axis.title.y = element_text(vjust= 1),
      # Change the distace of lables from x axis
      axis.text.x = element_text(vjust= -1),
      # Change the distace of lables from y axis
      axis.text.y = element_text(hjust= 2),
      # Remove the legend
      legend.position = "none",
      # Change the margin to make sure x axis labels aren't cut-off
      plot.margin = margin(10, 30, 10, 10)
    ) +
    theme(plot.title = element_text(hjust = 0.5))

```

Every mutation that is picked up by `Lofreq` is also picked up by `Pysam`. Since `lofreq` will use some statisitical test aside from simply counting the occurence of every base each postion, it makes sense that this is more stringent. The question is why those mutations that aren't picked up by lofreq. 

## Individual Validaiton of Spike Mutations. 

I wanted to individually validate each Spike mutation to ensure that they are being annotated correctly, and there are no 'multi-hit' (multiple SNPs in the same codon) mutations that are not correctly annotated. 

```{r Validate Spike Mutatnts}

# == Position 9 == #
pysam.df %>% filter(PROT_POS == 9) 
# P9L, 200X+ coverage, one timepoint @143, looks good. 

# == Position 18 == #
pysam.df %>% filter(PROT_POS == 18) 
# L18V, only 3 supporting reads (DP = 3), @146, not so good - probably drop.

# == Position 141 == #
pysam.df %>% filter(PROT_POS == 141) 
# L141F and L141*, both have extremely low coverage (2-4), not so good - probably drop

# == Position 144 == #
pysam.df %>% filter(PROT_POS == 144) 
# Multiple variants, all with very low coverage (< 10 support reads over these sites) - probably drop.

# == Position 183 == # ** Change `*` to `Y` **
pysam.df %>% filter(PROT_POS == 183) %>% View()
# This one appears real, there is a small (1%) frequency stop codon mutation that is improperly annotated. because
# it is on the background of (CAG -> CAT, CAT -> TAT) it's not a stop, but actually a 'Y', Tyrosine.

# == Position 262 == #
pysam.df %>% filter(PROT_POS == 262) 
# A262S, looks legit. 

# == Position 393 == #
pysam.df %>% filter(PROT_POS == 393) 
# T393S, looks legit. 

# == Position 440 == #
pysam.df %>% filter(PROT_POS == 440) 
# N440D, looks legit. 

# == Position 478 == #
pysam.df %>% filter(PROT_POS == 478) 
# T478K, looks legit. 

# == Position 484 == #
pysam.df %>% filter(PROT_POS == 484) 
# E484A and E484K, looks legit. 

# == Position 486 == #
pysam.df %>% filter(PROT_POS == 486) %>%  View()
# E484I and E484L, looks legit. 

# == Position 489 == #
pysam.df %>% filter(PROT_POS == 489)
# Y489H, looks legit. 

# == Position 493 == #
pysam.df %>% filter(PROT_POS == 493) %>% View()
# Q493K, looks legit. 

# == Position 494 == #
pysam.df %>% filter(PROT_POS == 494) 
# S494P, looks legit. 

# == Position 501 == #
pysam.df %>% filter(PROT_POS == 501)
# N501Y, looks legit. 

# == Position 837 == #
pysam.df %>% filter(PROT_POS == 837) %>% View()
# Y837H, looks legit. 

# == Position 870 == #
pysam.df %>% filter(PROT_POS == 870) %>% View()
# I870V, looks legit. 

# == Position 1020 == #
pysam.df %>% filter(PROT_POS == 1020) %>% View()
# A1020S, looks legit. 

# In summary, a depth cut-off of ~100 should remove the trouble maker positions where read coverage is too low (i.e., 18, 141, 144, and possibly 183)
```

After investigating every single postion and the discrepencies with `lofreq`, here is the pysam plot with a depth cutoff

```{r Variant Timecourse Spike - pysam > 100X, message=F, warning=F, echo=F, fig.width=15, fig.height=8}
## ==== Read in the pysam/pileup data ==== ##
pysam.df = read_csv(pysam.data)

## ==== Get a list of the amino acid positions to plot ==== ##
sites.to.plot = pysam.df %>% 
  # Must have at least one missense mutation
  filter(MISSENSE == "TRUE") %>% 
  # At least one mutation over 10%
  filter(AF >= 0.1) %>% 
  # Must have > 100X Coverage
  filter(DP >= 100) %>% 
  # Get a list of the relevant protein positions
  pull(PROT_POS) %>% 
  # only the unqiue sites are necessary
  unique()

## ==== Further filtering to remove sites that don't change over the interval ==== ##
figure.df = pysam.df[which(pysam.df$PROT_POS %in% sites.to.plot),] %>% 
  # Get only the missense mutations
  filter(MISSENSE == "TRUE") %>% 
  # Drop the extra columns 
  select(!c("ACCESSION", "MISSENSE", "DP", "CONS", "SNP")) %>% 
  # Make a new column for pivot 
  mutate(DAY = paste0("DAY_", DAY)) %>% 
  # Pivot wider to have information for every single time point
  pivot_wider(names_from = DAY, values_from = AF, values_fill = 0) %>%  
  # Pivot longer to tidy version. 
  pivot_longer(starts_with("DAY_"), names_to = "DAY", values_to = "AF") %>% 
  # Split the day back into a number 
  mutate(DAY = as.integer(str_replace_all(DAY, "DAY_", ""))) %>% 
  # Add the color information for plotting
  mutate(COLOR_SCHEME = case_when(AA_CHANGE == "N440D" ~ "ESCAPE",
                                  AA_CHANGE == "F486I" ~ "ESCAPE",
                                  AA_CHANGE == "Q493K" ~ "ESCAPE",
                                  AA_CHANGE == "E484A" ~ "ESCAPE",
                                  AA_CHANGE == "Y489H" ~ "ESCAPE",
                                  TRUE ~ "NON-ESCAPE"))


## ==== Plot Figure ==== ##
y.pos = 0
point.size = 1
adj = 0.05
tick.color = "#bcbd22"

figure.df %>% 
  # Remove D614G which is present in all timepoints.
  filter(AA_CHANGE != "D614G") %>%
  # Add "residue" to the name for plotting
  mutate(RESIDUE = fct_reorder(paste("Residue", PROT_POS, sep = " "), PROT_POS, .desc = F)) %>% 
  ## ========= Plot the figure ========= ##
  
  ggplot(aes(x = DAY, y = AF, fill = COLOR_SCHEME, group = MUT_AA)) +
    # Annotation to represent the WT allele frequency
    annotate(geom = "rect", xmin = 18, xmax = 152, ymin=0, ymax=1, fill = "#7f7f7f") +
    geom_area() +
    #geom_point(col ="black") +
    facet_wrap(~RESIDUE, ncol = 5) +
    ylab("mutation frequency") + 
    xlab("days after diagnosis") + 
    ggtitle("Psyam") +
    scale_fill_manual(values = tab_colors) +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(-.05,1) , expand = c(0,0)) +
    scale_x_continuous(limits = c(18,152) , expand = c(0,0)) +
    ### ==== Segments/Point Annotations X-Axis ==== ###
    annotate(geom = "segment", x = 145, y = 0 ,  xend = 145, yend = 1, size = 1, col = "black", linetype = 4) +
    annotate(geom = "segment", x = 18, y = 0 ,  xend = 152, yend = 0, size = 1, col = "black", linetype = 1) +
  
    annotate(geom = "segment", x = 18, y = y.pos ,  xend = 18, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 25, y = y.pos ,  xend = 25, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 75, y = y.pos ,  xend = 75, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 81, y = y.pos ,  xend = 81, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 128, y = y.pos ,  xend = 128, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 130, y = y.pos ,  xend = 130, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 143, y = y.pos ,  xend = 143, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 146, y = y.pos ,  xend = 146, yend = y.pos+adj, size = 1, col = tick.color) +
    annotate(geom = "segment", x = 152, y = y.pos ,  xend = 152, yend = y.pos+adj, size = 1, col = tick.color) +
    ### ==== Segments/Point Annotations X-Axis ==== ###
    
    ### ==== Themes for plot aestetics ==== ###
  
  
    theme_classic() +
    theme(
      # Set the appropriate font size
      text=element_text(size=20,  family="Helvetica"),
      # Change x axis title position
      axis.title.x = element_text(vjust= -0.5),
      # Change y axis title position
      axis.title.y = element_text(vjust= 1),
      # Change the distace of lables from x axis
      axis.text.x = element_text(vjust= -1),
      # Change the distace of lables from y axis
      axis.text.y = element_text(hjust= 2),
      # Remove the legend
      legend.position = "none",
      # Change the margin to make sure x axis labels aren't cut-off
      plot.margin = margin(10, 30, 10, 10)
    ) +
    theme(plot.title = element_text(hjust = 0.5))

```









